pkg/filewatch/filewatch.go

```diff
@@ -0,0 +1,118 @@
+package filewatch
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type Watch interface {
+	Add(path string, callback func()) error
+	Run(done <-chan struct{}) error
+	IsRunning() bool
+}
+
+func New() Watch {
+	return &watch{
+		callbacks: make(map[string]func()),
+	}
+}
+
+type watch struct {
+	lock      sync.Mutex
+	callbacks map[string]func()
+	running   atomic.Bool
+}
+
+var _ Watch = &watch{}
+
+func (w *watch) Add(path string, callback func()) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	if w.running.Load() {
+		return fmt.Errorf("cannot add to a running watch")
+	}
+
+	w.callbacks[path] = callback
+	return nil
+}
+
+func (w *watch) Run(done <-chan struct{}) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("could not create fsnotify.Watcher: %w", err)
+	}
+	// watcher.Close() never returns an error
+	defer func() { _ = watcher.Close() }()
+
+	func() {
+		// Before setting running to true, we need to acquire the lock,
+		// because Add() method may be running concurrently.
+		w.lock.Lock()
+		defer w.lock.Unlock()
+		w.running.Store(true)
+	}()
```

Comment:
Apologizes for my ignorance, why is this chunk of code inside a `func() {}`?

Reply:
```go
w.lock.Lock()
defer w.lock.Unlock()
```
Is a common pattern in golang. We use `defer` for unlock, so that we are sure that it's unlocked even on panic.

Here I use an anonymous function call to limit the scope of the critical section. So that `defer` is called after `w.running.Store(true)`. This is another pattern.

In this simple case it could have been:
```go
w.lock.Lock()
w.running.Store(true)
w.lock.Unlock()
```

Reply:
Cool! Thanks for the explanation! :blush: